cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandler.java

```diff
@@ -378,6 +382,70 @@ public void updateApp(FullHttpRequest request, HttpResponder responder,
     }
   }
 
+  /**
+   * upgrades an existing application.
+   */
+  @POST
+  @Path("/apps/{app-id}/upgrade")
+  @AuditPolicy(AuditDetail.REQUEST_BODY)
+  public void upgradeApplication(HttpRequest request, HttpResponder responder,
+                                 @PathParam("namespace-id") final String namespaceId,
+                                 @PathParam("app-id") final String appName)
+      throws NotFoundException, BadRequestException, UnauthorizedException, IOException {
+
+    ApplicationId appId = validateApplicationId(namespaceId, appName);
+
+    try {
+      ApplicationUpdateDetails detail =
+          applicationLifecycleService.upgradeApplication(appId, createProgramTerminator());
+      responder.sendJson(HttpResponseStatus.OK, GSON.toJson(detail));
+    } catch (Exception e) {
+      LOG.error("Upgrade application failure", e);
+      ApplicationUpdateDetails detail = new ApplicationUpdateDetails(
+          new ServiceException("Upgrade failed due to internal error.", null,
+                               HttpResponseStatus.INTERNAL_SERVER_ERROR));
+      responder.sendJson(HttpResponseStatus.OK, GSON.toJson(detail));
+    }
+  }
+
+  /**
+   * Upgrades a lis of existing application to use latest version of application artifact and plugin artifacts.
+   *
+   * <pre>
+   * {@code
+   * [
+   *   {"appId":"XYZ"},
+   *   {"appId":"ABC"},
+   *   {"appId":"FOO"},
+   * ]
+   * }
+   * </pre>
+   *
+   * The response will be map of Application ID to {@link ApplicationUpdateDetails} object, which either indicates a
```

Comment:
Don't use map as response for batch request. Response with array of objects. This is better to preserve the same ordering as the request.